src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs

```diff
@@ -0,0 +1,193 @@
+
+        // Rent the reusable ThreadPoolValueTaskSource, or create a new one to use if we couldn't get one (which
+        // should only happen on first use or if the SafeFileHandle is being used concurrently).
+        internal ThreadPoolValueTaskSource GetThreadPoolValueTaskSource() =>
+            Interlocked.Exchange(ref _reusableThreadPoolValueTaskSource, null) ?? new ThreadPoolValueTaskSource(this);
+
+        private void TryToReuse(ThreadPoolValueTaskSource source)
+        {
+            Interlocked.CompareExchange(ref _reusableThreadPoolValueTaskSource, source, null);
+        }
+
+        /// <summary>
+        /// A reusable <see cref="IValueTaskSource"/> implementation that
+        /// queues asynchronous <see cref="RandomAccess"/> operations to
+        /// be completed synchronously on the thread pool.
+        /// </summary>
+        internal sealed class ThreadPoolValueTaskSource : IThreadPoolWorkItem, IValueTaskSource<int>, IValueTaskSource<long>
+        {
+            private ManualResetValueTaskSourceCore<long> _source;
+            private readonly SafeFileHandle _fileHandle;
+            private Operation _operation = Operation.None;
+
+            private const bool PreferLocal = true;
+
+            // These fields store the parameters for the operation.
+            // The first two are common for all kinds of operations.
+            private long _fileOffset;
+            private CancellationToken _cancellationToken;
+            // Used by simple reads and writes. Will be unsafely cast to a memory when performing a read.
+            private ReadOnlyMemory<byte> _singleSegment;
+            // Used by vectored reads and writes. Is an IReadOnlyList of either Memory or ReadOnlyMemory of bytes.
+            private object? _multiSegmentCollection;
+
+            internal ThreadPoolValueTaskSource(SafeFileHandle fileHandle)
+            {
+                _fileHandle = fileHandle;
+            }
+
+            private long GetResultAndRelease(short token)
+            {
+                try
+                {
+                    return _source.GetResult(token);
+                }
+                finally
+                {
+                    _source.Reset();
+                    _fileHandle.TryToReuse(this);
+                }
+            }
+
+            public ValueTaskSourceStatus GetStatus(short token) => _source.GetStatus(token);
+            public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) =>
+                _source.OnCompleted(continuation, state, token, flags);
+            int IValueTaskSource<int>.GetResult(short token) => (int) GetResultAndRelease(token);
+            long IValueTaskSource<long>.GetResult(short token) => GetResultAndRelease(token);
+
+            void IThreadPoolWorkItem.Execute()
```

Comment:
I just noticed you're not flowing ExecutionContext.  I think we need to, in particular so that things like WindowsIdentity are properly restored; otherwise, this could lead to I/O failures due to permissions and the like.